### DISCUSSION TITLE
how to init workspace with "EMBEDDED"

### DISCUSSION THREAD

**Discussion by Shaquille-Wu** [osqp/osqp#554] [Embedded Optimization]
Hello, every one
I compiled osqp with "EMBEDDED",
but, I found I cannot init the "OSQPWorkspace", because, "osqp_setup" and "osqp_cleanup" just be compiled if "EMBEDDED" is false
the principle of osqp is very difficult for me, so, I want to know how to init the "OSQPWorkspace" with "EMBEDDED"?

**Comment by imciner2** [comment]
Hi @Shaquille-Wu, sorry for not replying to this earlier. The general setup for OSQP with the `EMBEDDED` flag is that the OSQP workspace and data should have been either previously manually allocated (not recommended), or generated using the code generation features in the MATLAB and Python interfaces. The purpose of the EMBEDDED mode is to remove all memory allocation to prevent possible errors, and ensure that small embedded devices don't run out of memory at run-time.

To use code generation, take a look at the documentation page (https://osqp.org/docs/release-0.6.3/codegen/index.html). It will require setting up your problem in either MATLAB or Python, and then calling the specific codegen function. That will generate a C file and a header file, which can then be added to your project. Then you just need to use `osqp_solve` with the variables in the generated header, and you can solve the optimization problem.
